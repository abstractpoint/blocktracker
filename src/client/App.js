import React, { useEffect, useState } from 'react';
import { ThemeProvider } from 'styled-components';
import {
  BrowserRouter as Router, Switch, Route,
} from 'react-router-dom';
import themeObj from './theme';
import { getLatestBlocksViaService } from './services/ethereum';
import Home from './scenes/home';
import Details from './scenes/details';
import GlobalStyles from './components/global-styles';
import { getAverageTimePerBlock, getAverageTxPerBlock } from './utils';

function App() {
  const [blocks, setBlocks] = useState([]);
  const [blockStats, setBlockStats] = useState({
    blockAverageSeconds: undefined,
    blockAverageTxNo: undefined,
  });

  const loadInitialBlocks = () => {
    const untilBlock = blocks.length > 0 && blocks[0].number;
    getLatestBlocksViaService(untilBlock).then(newBlocks => {
      if (newBlocks.length > 0) {
        const combinedBlocks = [
          ...newBlocks,
          ...blocks.slice(0, 19),
        ];

        setBlocks(combinedBlocks);
        setBlockStats({
          blockAverageSeconds: getAverageTimePerBlock(combinedBlocks),
          blockAverageTxNo: getAverageTxPerBlock(combinedBlocks),
        });
      }
    });
  };

  useEffect(loadInitialBlocks, []);
  useEffect(() => {
    const interval = setInterval(loadInitialBlocks, 4000);
    return () => clearInterval(interval);
  }, [blocks]);

  return (
    <ThemeProvider theme={themeObj}>
      <GlobalStyles />
      <Router>
        <Switch>
          <Route
            path="/block/:number"
            render={
              () => <Details blocks={blocks} />
            }
          />
          <Route path="/" render={() => <Home blocks={blocks} blockStats={blockStats} />} />
        </Switch>
      </Router>
    </ThemeProvider>
  );
}

export default App;
