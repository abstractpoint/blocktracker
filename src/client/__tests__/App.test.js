import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import App from '../App';

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

describe('App', () => {
  test('Should render without error', async () => {
    act(() => {
      ReactDOM.render(<App />, container);
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
