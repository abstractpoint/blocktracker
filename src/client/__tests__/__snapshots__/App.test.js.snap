// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render without error 1`] = `
"<div class=\\"sc-bwzfXH fXkOtp\\">
  <div class=\\"sc-htpNat dURhdG\\"><img src=\\"\\" alt=\\"whale logo (credit IFAW)\\" class=\\"sc-bwzfXH gLvYDQ\\">
    <div class=\\"sc-htoDjs kBpCXH\\">
      <div class=\\"sc-bwzfXH sc-dnqmqq elwQke\\">
        <div class=\\"sc-bwzfXH sc-iwsKbI fEddsE\\">
          <h2>Stats</h2>
        </div>
        <div class=\\"sc-bwzfXH sc-iwsKbI fEddsE\\"></div>
      </div>
      <div class=\\"sc-bwzfXH sc-dnqmqq gWDzYm\\"></div>
    </div>
  </div>
</div>"
`;
